File: app/src/main/java/com/zacharee1/systemuituner/activites/instructions/InstructionsActivity.java
package com.zacharee1.systemuituner.activites.instructions;

import android.content.Intent;
import android.support.annotation.ColorInt;
import android.support.annotation.IdRes;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.text.Html;
import android.text.Spanned;
import android.text.method.LinkMovementMethod;
import android.util.Log;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.github.paolorotolo.appintro.AppIntro2;
import com.github.paolorotolo.appintro.ISlideBackgroundColorHolder;
import com.zacharee1.systemuituner.R;
import com.zacharee1.systemuituner.util.Utils;

import java.util.ArrayList;

public class InstructionsActivity extends AppIntro2 {
    public static final String ARG_TITLE = "title";
    public static final String ARG_DESC = "desc";
    public static final String ARG_LAYOUT = "layout";
    public static final String ARG_BG_COLOR = "bg_color";
    public static final String ARG_TITLE_COLOR = "title_color";
    public static final String ARG_DESC_COLOR = "desc_color";
    public static final String ARG_COMMANDS = "commands";

    private static final String WINDOWS = "windows";
    private static final String MAC = "mac";
    private static final String UBUNTU = "ubuntu";
    private static final String LINUX = "linux";
    private static final String FEDORA = "fedora";

    private static final String PREFIX = "adb shell pm grant com.zacharee1.systemuituner ";

    private Instructions mSelector;
    private Instructions mInitial;
    private Commands mCommands;

    private Instructions mInstructions;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setTheme(R.style.AppTheme_Intro);
        super.onCreate(savedInstanceState);

        mInstructions = Instructions.newInstance(getResources().getString(R.string.windows_setup),
                getResources().getString(R.string.on_computer),
                R.layout.fragment_adb_windows,
                getResources().getColor(R.color.intro_1, null));

        mSelector = Instructions.newInstance(getResources().getString(R.string.choose_your_weapon),
                getResources().getString(R.string.which_os),
                R.layout.fragment_adb_select,
                getResources().getColor(R.color.intro_5, null));

        mInitial = Instructions.newInstance(getResources().getString(R.string.initial_setup),
                getResources().getString(R.string.on_device),
                R.layout.fragment_adb_initial,
                getResources().getColor(R.color.intro_2, null));

        Intent intent = getIntent();
        Bundle extras = intent.getExtras();
        ArrayList<String> cmds = extras.getStringArrayList(ARG_COMMANDS);

        mCommands = Commands.newInstance(getResources().getString(R.string.run_commands),
                getResources().getString(R.string.run_on_computer),
                getResources().getColor(R.color.intro_4, null),
                cmds);

        addSlide(mSelector);
        addSlide(mInitial);
        addSlide(mInstructions);
        addSlide(mCommands);

        pager.setOffscreenPageLimit(300);

        backButton.setVisibility(View.GONE);
        skipButtonEnabled = false;
        doneButton.setVisibility(View.GONE);
//        showPagerIndicator(false);

        setColorTransitionsEnabled(true);
    }

    @Override
    public void onSlideChanged(@Nullable Fragment oldFragment, @Nullable Fragment newFragment) {
        int oldIndex = fragments.indexOf(oldFragment);
        int newIndex = fragments.indexOf(newFragment);

        if (newIndex == fragments.size() - 1 || newIndex == -1) {
            nextButton.setVisibility(View.GONE);
            doneButton.setVisibility(View.VISIBLE);
        } else if (newIndex > 0) {
            backButton.setVisibility(View.VISIBLE);
            nextButton.setVisibility(View.VISIBLE);
            doneButton.setVisibility(View.GONE);
        } else {
            backButton.setVisibility(View.GONE);
            doneButton.setVisibility(View.GONE);
            nextButton.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void onDonePressed(Fragment currentFragment) {
        finish();
    }

    private void replaceContentById(@LayoutRes int layout) {
        mInstructions.setInternalLayout(layout);
    }

    private void setInstructionsTitle(String title) {
        mInstructions.setTitle(title);
    }

    public void chooseWindows(View v) {
        replaceContentById(R.layout.fragment_adb_windows);
        setInstructionsTitle(getResources().getString(R.string.windows_setup));
    }

    public void chooseMac(View v) {
        replaceContentById(R.layout.fragment_adb_mac);
        setInstructionsTitle(getResources().getString(R.string.mac_setup));
    }

    public void chooseUbuntu(View v) {
        replaceContentById(R.layout.fragment_adb_ubuntu);
        setInstructionsTitle(getResources().getString(R.string.ubuntu_setup));
    }

    public void chooseLinux(View v) {
        replaceContentById(R.layout.fragment_adb_linux);
        setInstructionsTitle(getResources().getString(R.string.linux_setup));
    }

    public void chooseFedora(View v) {
        replaceContentById(R.layout.fragment_adb_fedora);
        setInstructionsTitle(getResources().getString(R.string.fedora_setup));
    }

    public static class Instructions extends Fragment implements ISlideBackgroundColorHolder {
        protected View mView;

        public static Instructions newInstance(CharSequence title, CharSequence description,
                                           @LayoutRes int layoutId,
                                           @ColorInt int bgColor) {
            return newInstance(title, description, layoutId, bgColor, 0, 0);
        }

        public static Instructions newInstance(CharSequence title, CharSequence description,
                                           @LayoutRes int layoutId, @ColorInt int bgColor,
                                           @ColorInt int titleColor, @ColorInt int descColor) {
            Instructions slide = new Instructions();
            Bundle args = new Bundle();
            args.putString(ARG_TITLE, title.toString());
            args.putString(ARG_DESC, description.toString());
            args.putInt(ARG_LAYOUT, layoutId);
            args.putInt(ARG_BG_COLOR, bgColor);
            args.putInt(ARG_TITLE_COLOR, titleColor);
            args.putInt(ARG_DESC_COLOR, descColor);
            slide.setArguments(args);

            return slide;
        }

        @Nullable
        @Override
        public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
            mView = inflater.inflate(getLayoutId(), container, false);

            Log.e("onCreateView()", getArguments().getString(ARG_TITLE));

            View main = mView.findViewById(R.id.main);
            TextView title = mView.findViewById(R.id.title);
            TextView desc = mView.findViewById(R.id.description);

            Bundle args = getArguments();

            main.setBackgroundColor(args.getInt(ARG_BG_COLOR));
            title.setText(args.getString(ARG_TITLE));
            desc.setText(args.getString(ARG_DESC));

            if (args.getInt(ARG_LAYOUT) != 0 && args.getInt(ARG_LAYOUT) != -1) {
                setInternalLayout(args.getInt(ARG_LAYOUT));
            }

            return mView;
        }

        public <T extends View> T findViewById(@IdRes int id) {
            return mView.findViewById(id);
        }

        public int getLayoutId() {
            return R.layout.fragment_intro_custom_center;
        }

        public void setTitle(String title) {
            TextView textView = mView.findViewById(R.id.title);
            textView.setText(formatText(title));
        }

        public void setInternalLayout(@LayoutRes int layoutId) {
            LinearLayout holder = mView.findViewById(R.id.custom_layout_holder);
            holder.removeAllViews();

            ViewGroup viewParent = (ViewGroup) View.inflate(getActivity(), layoutId, null);
            ViewGroup viewChild = (ViewGroup) viewParent.getChildAt(0);

            ViewGroup view = viewChild != null ? viewChild : viewParent;

            for (int i = 0; i < view.getChildCount(); i++) {
                View v = view.getChildAt(i);

                if (v instanceof TextView) {
                    ((TextView) v).setText(formatText(((TextView) v).getText().toString()));
                    ((TextView) v).setLinksClickable(true);
                    ((TextView) v).setMovementMethod(LinkMovementMethod.getInstance());
                    ((TextView) v).setLinkTextColor(getResources().getColorStateList(R.color.white, null));
                    ((TextView) v).setTextColor(getResources().getColorStateList(R.color.white, null));
                }
            }

            holder.addView(viewParent);
        }

        @Override
        public int getDefaultBackgroundColor() {
            return getArguments().getInt(ARG_BG_COLOR);
        }

        @Override
        public void setBackgroundColor(int backgroundColor) {
            mView.setBackgroundColor(backgroundColor);
        }

        private Spanned formatText(String text) {
            return Html.fromHtml(text);
        }
    }

    public static class Commands extends Instructions {
        public static Commands newInstance(CharSequence title, CharSequence description,
                                               @ColorInt int bgColor,
                                               ArrayList<String> commands) {
            return newInstance(title, description, bgColor, 0, 0, commands);
        }

        public static Commands newInstance(CharSequence title, CharSequence description,
                                               @ColorInt int bgColor,
                                               @ColorInt int titleColor, @ColorInt int descColor,
                                               ArrayList<String> commands) {
            Commands slide = new Commands();
            Bundle args = new Bundle();
            args.putString(ARG_TITLE, title.toString());
            args.putString(ARG_DESC, description.toString());
            args.putInt(ARG_BG_COLOR, bgColor);
            args.putInt(ARG_TITLE_COLOR, titleColor);
            args.putInt(ARG_DESC_COLOR, descColor);
            args.putStringArrayList(ARG_COMMANDS, commands);
            slide.setArguments(args);

            return slide;
        }

        @Nullable
        @Override
        public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
            View view = super.onCreateView(inflater, container, savedInstanceState);

            ArrayList<String> commands = getArguments().getStringArrayList(ARG_COMMANDS);
            LinearLayout holder = view.findViewById(R.id.custom_layout_holder);

            for (String command : commands) {
                TextView textView = new TextView(getActivity());
                textView.setGravity(Gravity.CENTER_HORIZONTAL | Gravity.CENTER_VERTICAL);
                textView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 18);
                textView.setTextIsSelectable(true);
                textView.setText(PREFIX.concat(command));
                textView.setPadding(
                        0,
                        (int)Utils.pxToDp(getActivity(), 8),
                        0,
                        0
                );

                holder.addView(textView);
            }

            return view;
        }
    }
}

File: app/src/main/java/com/zacharee1/systemuituner/activites/instructions/InstructionsActivity.kt
package com.zacharee1.systemuituner.activites.instructions

import android.animation.Animator
import android.annotation.SuppressLint
import android.os.Build
import android.os.Bundle
import android.text.Html
import android.text.Spanned
import android.text.method.LinkMovementMethod
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.annotation.ColorInt
import androidx.annotation.IdRes
import androidx.annotation.LayoutRes
import androidx.core.text.HtmlCompat
import androidx.fragment.app.Fragment
import com.github.paolorotolo.appintro.AppIntro2
import com.github.paolorotolo.appintro.ISlideBackgroundColorHolder
import com.github.paolorotolo.appintro.ISlidePolicy
import com.zacharee1.systemuituner.R
import kotlinx.android.synthetic.main.command_box.view.*
import java.util.*

class InstructionsActivity : AppIntro2() {

    private lateinit var initial: Instructions
    private lateinit var commands: Commands
    private lateinit var instructions: OSInstructions

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        instructions = OSInstructions.newInstance(resources.getString(R.string.choose_your_weapon),
                resources.getString(R.string.which_os),
                R.layout.fragment_adb_select,
                resources.getColor(R.color.intro_5, null))

        initial = Instructions.newInstance(resources.getString(R.string.initial_setup),
                resources.getString(R.string.on_device),
                R.layout.fragment_adb_initial,
                resources.getColor(R.color.intro_2, null))

        val intent = intent
        if (intent == null) {
            finish()
            return
        }

        val extras = intent.extras
        if (extras == null) {
            finish()
            return
        }

        val cmds = extras.getStringArrayList(ARG_COMMANDS)

        commands = Commands.newInstance(resources.getString(R.string.run_commands),
                resources.getString(R.string.run_on_computer),
                resources.getColor(R.color.intro_6, null),
                cmds)

        addSlide(initial)
        addSlide(instructions)
        addSlide(commands)

        pager.offscreenPageLimit = 300

        backButton.visibility = View.GONE
        skipButtonEnabled = false
        doneButton.visibility = View.GONE

        setColorTransitionsEnabled(true)
    }

    override fun onSlideChanged(oldFragment: Fragment?, newFragment: Fragment?) {
        val newIndex = fragments.indexOf(newFragment)

        if (newIndex == fragments.size - 1 || newIndex == -1) {
            nextButton.visibility = View.GONE
            doneButton.visibility = View.VISIBLE
        } else if (newIndex > 0) {
            backButton.visibility = View.VISIBLE
            nextButton.visibility = View.VISIBLE
            doneButton.visibility = View.GONE
        } else {
            backButton.visibility = View.GONE
            doneButton.visibility = View.GONE
            nextButton.visibility = View.VISIBLE
        }
    }

    override fun onDonePressed(currentFragment: Fragment?) {
        finish()
    }

    open class OSInstructions: Instructions(), ISlidePolicy {
        companion object {
            @JvmOverloads
            fun newInstance(title: CharSequence, description: CharSequence,
                            @LayoutRes layoutId: Int, @ColorInt bgColor: Int,
                            @ColorInt titleColor: Int = 0, @ColorInt descColor: Int = 0): OSInstructions {
                val slide = OSInstructions()
                val args = Bundle()
                args.putString(ARG_TITLE, title.toString())
                args.putString(ARG_DESC, description.toString())
                args.putInt(ARG_LAYOUT, layoutId)
                args.putInt(ARG_BG_COLOR, bgColor)
                args.putInt(ARG_TITLE_COLOR, titleColor)
                args.putInt(ARG_DESC_COLOR, descColor)
                slide.arguments = args

                return slide
            }
        }

        private val selection by lazy { findViewById<Button>(R.id.change_selection) }

        private var hasSelectedAnOs = false
        private var isOnOsSelectionScreen = true

        override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
            super.onViewCreated(view, savedInstanceState)

            addSelectorButton()
            setSelectionListeners()
        }

        override fun isPolicyRespected(): Boolean {
            return !isOnOsSelectionScreen || hasSelectedAnOs
        }

        override fun onUserIllegallyRequestedNextPage() {
            Toast.makeText(context, R.string.select_os, Toast.LENGTH_SHORT).show()
        }

        private fun setSelectionListeners() {
            selection?.visibility = View.GONE

            val windows = findViewById<RadioButton>(R.id.choose_windows)
            val mac = findViewById<RadioButton>(R.id.choose_mac)
            val ubuntu = findViewById<RadioButton>(R.id.choose_ubuntu)
            val fedora = findViewById<RadioButton>(R.id.choose_fedora)
            val linux = findViewById<RadioButton>(R.id.choose_linux)

            val clickListener = View.OnClickListener {
                var title = 0
                var layout = 0

                when (it.id) {
                    R.id.choose_windows -> {
                        title = R.string.windows_setup
                        layout = R.layout.fragment_adb_windows
                    }

                    R.id.choose_mac -> {
                        title = R.string.mac_setup
                        layout = R.layout.fragment_adb_mac
                    }

                    R.id.choose_ubuntu -> {
                        title = R.string.ubuntu_setup
                        layout = R.layout.fragment_adb_ubuntu
                    }

                    R.id.choose_fedora -> {
                        title = R.string.fedora_setup
                        layout = R.layout.fragment_adb_fedora
                    }

                    R.id.choose_linux -> {
                        title = R.string.linux_setup
                        layout = R.layout.fragment_adb_linux
                    }
                }

                animateChange(layout, title, R.string.on_computer,
                        Runnable { selection?.visibility = View.VISIBLE })
                isOnOsSelectionScreen = false
                hasSelectedAnOs = true
            }

            windows?.setOnClickListener(clickListener)
            mac?.setOnClickListener(clickListener)
            ubuntu?.setOnClickListener(clickListener)
            fedora?.setOnClickListener(clickListener)
            linux?.setOnClickListener(clickListener)
        }

        private fun addSelectorButton() {
            selection?.setOnClickListener {
                animateChange(R.layout.fragment_adb_select, R.string.choose_your_weapon, R.string.which_os, Runnable { setSelectionListeners() })
                isOnOsSelectionScreen = true
                hasSelectedAnOs = false
            }
        }

        private fun animateChange(targetLayout: Int, targetTitle: Int, targetDesc: Int, runnable: Runnable) {
            view?.findViewById<LinearLayout>(R.id.animation_dummy)?.animate()?.let {
                it.setListener(object : Animator.AnimatorListener {
                    override fun onAnimationEnd(animation: Animator?) {
                        setInternalLayout(View.inflate(context ?: return, targetLayout, null) as ViewGroup)
                        setTitle(resources.getString(targetTitle))
                        setDescription(resources.getString(targetDesc))
                        runnable.run()

                        it.setListener(null)
                        it.alpha(1.0f).setDuration(500L).start()
                    }

                    override fun onAnimationCancel(animation: Animator?) {}

                    override fun onAnimationRepeat(animation: Animator?) {}

                    override fun onAnimationStart(animation: Animator?) {}
                })

                it.alpha(0.0f).setDuration(500L).start()
            }
        }
    }

    open class Instructions : Fragment(), ISlideBackgroundColorHolder {
        private val layoutId: Int = R.layout.fragment_intro_custom_center

        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            return inflater.inflate(layoutId, container, false)
        }

        override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
            super.onViewCreated(view, savedInstanceState)

            val main = view.findViewById<View>(R.id.main)
            val title = view.findViewById<TextView>(R.id.title)
            val desc = view.findViewById<TextView>(R.id.description)

            val args = arguments

            main.setBackgroundColor(args!!.getInt(ARG_BG_COLOR))
            title.text = args.getString(ARG_TITLE)
            desc.text = args.getString(ARG_DESC)

            val argLayout = args.getInt(ARG_LAYOUT)

            if (argLayout > 0) {
                val group = View.inflate(activity, argLayout, null) as ViewGroup
                setInternalLayout(group)
            }
        }

        fun <T : View> findViewById(@IdRes id: Int): T? {
            return view?.findViewById(id)
        }

        fun setTitle(title: String) {
            findViewById<TextView>(R.id.title)?.text = formatText(title)
        }

        fun setDescription(description: String) {
            findViewById<TextView>(R.id.description)?.text = formatText(description)
        }

        fun setInternalLayout(group: ViewGroup) {
            val holder = view?.findViewById<LinearLayout>(R.id.custom_layout_holder)
            holder?.removeAllViews()

            val viewChild = group.getChildAt(0) as ViewGroup

            for (i in 0 until viewChild.childCount) {
                val v = viewChild.getChildAt(i)

                if (v is TextView && v !is Button) {
                    v.text = formatText(v.text.toString())
                    v.linksClickable = true
                    v.movementMethod = LinkMovementMethod.getInstance()
                    v.setLinkTextColor(resources.getColorStateList(R.color.white, null))
                    v.setTextColor(resources.getColorStateList(R.color.white, null))
                }
            }

            holder?.addView(group)
        }

        override fun getDefaultBackgroundColor(): Int {
            return arguments!!.getInt(ARG_BG_COLOR)
        }

        override fun setBackgroundColor(backgroundColor: Int) {
            view!!.setBackgroundColor(backgroundColor)
        }

        private fun formatText(text: String): Spanned {
            return if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M) Html.fromHtml(text, 0) else HtmlCompat.fromHtml(text, 0)
        }

        companion object {
            @JvmOverloads
            fun newInstance(title: CharSequence, description: CharSequence,
                            @LayoutRes layoutId: Int, @ColorInt bgColor: Int,
                            @ColorInt titleColor: Int = 0, @ColorInt descColor: Int = 0): Instructions {
                val slide = Instructions()
                val args = Bundle()
                args.putString(ARG_TITLE, title.toString())
                args.putString(ARG_DESC, description.toString())
                args.putInt(ARG_LAYOUT, layoutId)
                args.putInt(ARG_BG_COLOR, bgColor)
                args.putInt(ARG_TITLE_COLOR, titleColor)
                args.putInt(ARG_DESC_COLOR, descColor)
                slide.arguments = args

                return slide
            }
        }
    }

    class Commands : Instructions() {

        @SuppressLint("SetTextI18n")
        override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
            super.onViewCreated(view, savedInstanceState)

            val commands = arguments?.getStringArrayList(ARG_COMMANDS) ?: return
            val holder = view.findViewById<LinearLayout>(R.id.custom_layout_holder)

            for (command in commands) {
                val commandBox = layoutInflater.inflate(R.layout.command_box, holder, false)

                commandBox.command.text = "adb shell pm grant ${view.context.packageName} $command"
                holder.addView(commandBox)
            }
        }

        companion object {
            fun newInstance(title: CharSequence, description: CharSequence,
                            @ColorInt bgColor: Int,
                            commands: ArrayList<String>?): Commands {
                return newInstance(title, description, bgColor, 0, 0, commands)
            }

            private fun newInstance(title: CharSequence, description: CharSequence,
                                    @ColorInt bgColor: Int,
                                    @ColorInt titleColor: Int, @ColorInt descColor: Int,
                                    commands: ArrayList<String>?): Commands {
                val slide = Commands()
                val args = Bundle()
                args.putString(ARG_TITLE, title.toString())
                args.putString(ARG_DESC, description.toString())
                args.putInt(ARG_BG_COLOR, bgColor)
                args.putInt(ARG_TITLE_COLOR, titleColor)
                args.putInt(ARG_DESC_COLOR, descColor)
                args.putStringArrayList(ARG_COMMANDS, commands)
                slide.arguments = args

                return slide
            }
        }
    }

    companion object {
        const val ARG_TITLE = "title"
        const val ARG_DESC = "desc"
        const val ARG_LAYOUT = "layout"
        const val ARG_BG_COLOR = "bg_color"
        const val ARG_TITLE_COLOR = "title_color"
        const val ARG_DESC_COLOR = "desc_color"
        const val ARG_COMMANDS = "commands"

        private const val PREFIX = "adb shell pm grant com.zacharee1.systemuituner "
    }
}
